### DISCUSSION TITLE
Dudas parcia Nefli

### DISCUSSION THREAD

**Discussion by Gruschka** [dds-jv/foro#169] [Consultas de objetos] [answered]
Buenas!

Estoy hace un buen rato ya con el [parcial Nefli](https://docs.google.com/document/d/1YYFk99LmQp1OrRuyhAagUVa4J8lwGkMHd_tbj5gHYk8/edit), dando vueltas hace horas sin llegar a buen puerto. Mis duda pasa principalmente sobre como usar el componente Reproductor.

En la premisa nos hablan de otro equipo que se encarga de crear el componente para reproducir las peliculas, quien nos pasó la siguiente interfaz:

<img width="331" alt="image" src="https://user-images.githubusercontent.com/11674324/89849775-9bc55b80-db5f-11ea-8bf9-0a07ed686a9d.png">

Me está costando mucho entender bien cómo usar este componente.  Por lo que veo en la interfaz, nos dan la pista de que 'el listener' debe implementar ciertos metodos, y que el Reproductor tiene métodos para suscribir a distintos objetos como listeners de los eventos play/stop/finished.

Por un lado, lo que pienso es que algún objeto tiene que 'usar' el Reproductor (es decir, llamar a los metodos play/stop). Yo entiendo que si bien hacemos mucho enfasis en clase que los roles del dominio no siempre se traducen en clases del sistema, no estaría mal en este caso tener una clase Usuario que llame al play() y el stop() del reproductor (despues incluso habla de distintos perfiles de Usuario) a riesgo de que esta clase se un middle man. Lo que me lleva a los listeners:

Lo lógico para mi es que los listeners sean las películas para que sepan qué hacer ante un onStop o un onFinished(). Ahora bien, lo que me tira abajo esta idea es el siguiente requerimiento:

> Una vez que se termina de ver una de las películas de la saga, tiene que reproducir la película que le sigue  hasta que no haya ninguna más.

Esto me da la pauta de que el objeto suscripto al onFinished tiene que poder conocer el Reproductor para pedirle, si corresponde, que haga play() con la siguiente pelicula de la saga. 

En mi solucion, es el Usuario quien tiene un reproductor. Sería válido hacer que las películas puedan decirle al Reproductor que hagan play de otra película? Cuál sería la relacion entre Usuario - Peliculas - Reproductor adecuada? Me hace ruido que ambos 'tengan' un reproductor. Estoy muy errado? Además no tengo en claro quién se sucribiría al 'onPlay'. El usuario no le encuentro sentido ya que, segun entiendo, es este objeto quien dispararía este evento. La película por otra parte, no hace nada en funcion a este evento (ya que la accion de 'reproducir' la pelicula la haria el reproductor).

Perdon por el wall of text, quería dejar en claro mi linea de razonamiento para ver qué tan lejos estoy de lo que se espera del ejercicio.

Mil gracias!

**Comment by gastonprieto** [comment, marked as answer]
> Me está costando mucho entender bien cómo usar este componente. Por lo que veo en la interfaz, nos dan la pista de que 'el listener' debe implementar ciertos metodos, y que el Reproductor tiene métodos para suscribir a distintos objetos como listeners de los eventos play/stop/finished.

Correcto

> Por un lado, lo que pienso es que algún objeto tiene que 'usar' el Reproductor (es decir, llamar a los metodos play/stop). Yo entiendo que si bien hacemos mucho enfasis en clase que los roles del dominio no siempre se traducen en clases del sistema, no estaría mal en este caso tener una clase Usuario que llame al play() y el stop() del reproductor (despues incluso habla de distintos perfiles de Usuario) a riesgo de que esta clase se un middle man. Lo que me lleva a los listeners:

Para mi lo clave acá es si tu clase usuario en el método que vos queres ponerle solo hace:

```
#Usuario>>play(reproductor)
reproductor.play(...)
```

Si hace solo eso entonces estariamos ante un middle man .

Igual tal vez la clase usuario existe, pero solo que acá no se usa.

> Lo lógico para mi es que los listeners sean las películas para que sepan qué hacer ante un onStop o un onFinished

Tiene sentido, es eso o es algo que conoce a la pelicula/serie.

> Esto me da la pauta de que el objeto suscripto al onFinished tiene que poder conocer el Reproductor para pedirle, si corresponde, que haga play() con la siguiente pelicula de la saga.

Si.

> Sería válido hacer que las películas puedan decirle al Reproductor que hagan play de otra película?

Si es lo que necesitas si 👍 

> Cuál sería la relacion entre Usuario - Peliculas - Reproductor adecuada? Me hace ruido que ambos 'tengan' un reproductor

Esto depende de tu solución. Pero comentario, existe otras formas de acceder al Reproductor aparte de tenerlo.
Pero si, ambos podrían conocer al reproductor, porque ambos le quieren dar play.

> Además no tengo en claro quién se sucribiría al 'onPlay'. El usuario no le encuentro sentido ya que, segun entiendo, es este objeto quien dispararía este evento. La película por otra parte, no hace nada en funcion a este evento (ya que la accion de 'reproducir' la pelicula la haria el reproductor).

Exacto, nadie lo usaría por eso mismo que comentas vos. No siempre se usan todos los mensajes de los componentes que usamos.

**Comment by Gruschka** [comment]
Gastón, mil gracias otra vez por la respuesta detallada! Clarísimo todo.
